app/src/main/java/com/ahugenb/hra/calculator/CalculatorView.kt

```diff
@@ -0,0 +1,162 @@
+                        .fillMaxWidth(),
+                )
+                Text(
+                    textAlign = TextAlign.Right,
+                    text = stringResource(ethanolId, pureEthanol),
+                    style = MaterialTheme.typography.h4,
+                    modifier = Modifier
+                        .padding(horizontal = 8.dp)
+                        .fillMaxWidth()
+                )
+            }
+        }
+        Row(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(start = 8.dp, end = 8.dp, bottom = 8.dp),
+            horizontalArrangement = Arrangement.Center
+        ) {
+            OutlinedTextField(
+                maxLines = 1,
+                modifier = Modifier
+                    .weight(1f, true)
+                    .padding(end = 8.dp),
+                value = volume.value,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal, imeAction = ImeAction.Next),
+                onValueChange = {
+                    if (it.acceptVolumeText()) {
+                        volume.value = it
+                        viewModel.updateCalculation(
+                            abv.value.smartToDouble(),
+                            volume.value.smartToDouble(),
+                            drinks.value.smartToDouble(),
+                            mlChecked.value
+                        )
+                    }
+                },
+                label = { Text(text = stringResource(labelId)) }
+            )
+            OutlinedTextField(
+                maxLines = 1,
+                modifier = Modifier
+                    .weight(1f, true)
+                    .padding(end = 8.dp),
+                value = abv.value,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal, imeAction = ImeAction.Next),
+                onValueChange = {
+                    if (it.acceptPercentText()) {
+                        abv.value = it
+                        viewModel.updateCalculation(
+                            abv.value.smartToDouble(),
+                            volume.value.smartToDouble(),
+                            drinks.value.smartToDouble(),
+                            mlChecked.value
+                        )
+                    }
+                },
+                label = { Text(text = "% ABV") }
+            )
+            OutlinedTextField(
+                maxLines = 1,
+                modifier = Modifier.weight(1f, true),
+                value = drinks.value,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal,
+                    imeAction = ImeAction.Done),
+                keyboardActions = KeyboardActions(onDone = { focusManager.clearFocus() }),
+                onValueChange = {
+                    if (it.acceptDrinksText()) {
+                        drinks.value = it
+                        viewModel.updateCalculation(
```

Comment:
Just feels like you shouldn't have to pass all the values to your viewmodel every time. Should be able to just update the value, and the viewmodel updates it's calculation, then the side effects of the calculation get rendered.